## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to download a movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [1]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [81]:
# Write your code for importing the csv file here
movies = pd.read_csv('Movies.csv', header = 0)
movies.head()

movies.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [23]:
# Write your code for inspection here
#movies.shape

#(movies.isna().sum() > 0).sum()

12

#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [32]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations

#movies.drop(columns = ['color', 'director_facebook_likes' , 'actor_1_facebook_likes' , 'actor_2_facebook_likes' , 'actor_3_facebook_likes' , 'actor_2_name' , 'cast_total_facebook_likes', 'actor_3_name', 'duration', 'facenumber_in_poster', 'content_rating', 'country', 'movie_imdb_link', 'aspect_ratio', 'plot_keywords'])

director_name  num_critic_for_reviews        gross  \
0         James Cameron                   723.0  760505847.0   
1        Gore Verbinski                   302.0  309404152.0   
2            Sam Mendes                   602.0  200074175.0   
3     Christopher Nolan                   813.0  448130642.0   
4        Andrew Stanton                   462.0   73058679.0   
...                 ...                     ...          ...   
3848      Shane Carruth                   143.0     424760.0   
3849   Neill Dela Llana                    35.0      70071.0   
3850   Robert Rodriguez                    56.0    2040920.0   
3851       Edward Burns                    14.0       4584.0   
3852           Jon Gunn                    43.0      85222.0   

                                   genres     actor_1_name  \
0         Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1                Action|Adventure|Fantasy      Johnny Depp   
2               Action|Adventure|Thriller  Christoph Waltz   
3                         Action|Thriller        Tom Hardy   
4                 Action|Adventure|Sci-Fi     Daryl Sabara   
...                                   ...              ...   
3848                Drama|Sci-Fi|Thriller    Shane Carruth   
3849                             Thriller      Ian Gamazon   
3850  Action|Crime|Drama|Romance|Thriller  Carlos Gallardo   
3851                         Comedy|Drama      Kerry Bishé   
3852                          Documentary      John August   

                                    movie_title  num_voted_users  \
0                                       Avatar            886204   
1     Pirates of the Caribbean: At World's End            471220   
2                                      Spectre            275868   
3                        The Dark Knight Rises           1144337   
4                                  John Carter            212204   
...                                         ...              ...   
3848                                    Primer             72639   
3849                                    Cavite               589   
3850                               El Mariachi             52055   
3851                                 Newlyweds              1338   
3852                         My Date with Drew              4285   

      num_user_for_reviews language       budget  title_year  imdb_score  \
0                   3054.0  English  237000000.0      2009.0         7.9   
1                   1238.0  English  300000000.0      2007.0         7.1   
2                    994.0  English  245000000.0      2015.0         6.8   
3                   2701.0  English  250000000.0      2012.0         8.5   
4                    738.0  English  263700000.0      2012.0         6.6   
...                    ...      ...          ...         ...         ...   
3848                 371.0  English       7000.0      2004.0         7.0   
3849                  35.0  English       7000.0      2005.0         6.3   
3850                 130.0  Spanish       7000.0      1992.0         6.9   
3851                  14.0  English       9000.0      2011.0         6.4   
3852                  84.0  English       1100.0      2004.0         6.6   

      movie_facebook_likes  
0                    33000  
1                        0  
2                    85000  
3                   164000  
4                    24000  
...                    ...  
3848                 19000  
3849                    74  
3850                     0  
3851                   413  
3852                   456  

[3853 rows x 13 columns]

#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'. 

In [57]:
# Write you code here
round(movies.isnull().sum() / len(movies.index) * 100, 2)


color                        0.05
director_name                0.00
num_critic_for_reviews       0.03
duration                     0.03
director_facebook_likes      0.00
actor_3_facebook_likes       0.16
actor_2_name                 0.03
actor_1_facebook_likes       0.00
gross                        0.00
genres                       0.00
actor_1_name                 0.00
movie_title                  0.00
num_voted_users              0.00
cast_total_facebook_likes    0.00
actor_3_name                 0.16
facenumber_in_poster         0.16
plot_keywords                0.78
movie_imdb_link              0.00
num_user_for_reviews         0.00
language                     0.08
country                      0.00
content_rating               1.25
budget                       0.00
title_year                   0.00
actor_2_facebook_likes       0.03
imdb_score                   0.00
aspect_ratio                 1.87
movie_facebook_likes         0.00
dtype: float64

#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [77]:
# Write your code for filling the NaN values in the 'language' column here
movies.loc[pd.isnull(movies['language']), ['language']] = 'English'

(movies.language == 'English').sum()

3674

#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
28    Color    Colin Trevorrow                   644.0     124.0   
25    Color      James Cameron                   315.0     194.0   
2704  Color       George Lucas                   282.0     125.0   
2748  Color   Steven Spielberg                   215.0     120.0   
16    Color        Joss Whedon                   703.0     173.0   
482   Color       Roger Allers                   186.0      73.0   
230   Color       George Lucas                   320.0     136.0   
64    Color  Christopher Nolan                   645.0     152.0   
419   Color          Gary Ross                   673.0     142.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
0                         0.0                   855.0   Joel David Moore   
28                      365.0                  1000.0         Judy Greer   
25                        0.0                   794.0       Kate Winslet   
2704                      0.0                   504.0      Peter Cushing   
2748                  14000.0                   548.0        Dee Wallace   
16                        0.0                 19000.0  Robert Downey Jr.   
482                      28.0                   847.0        Nathan Lane   
230                       0.0                  1000.0        Liam Neeson   
64                    22000.0                 11000.0       Heath Ledger   
419                     378.0                   575.0    Josh Hutcherson   

      actor_1_facebook_likes        gross  \
0                     1000.0  760505847.0   
28                    3000.0  652177271.0   
25                   29000.0  658672302.0   
2704                 11000.0  460935665.0   
2748                   861.0  434949459.0   
16                   26000.0  623279547.0   
482                   2000.0  422783777.0   
230                  20000.0  474544677.0   
64                   23000.0  533316061.0   
419                  34000.0  407999255.0   

                                        genres  ... language country  \
0              Action|Adventure|Fantasy|Sci-Fi  ...  English     USA   
28            Action|Adventure|Sci-Fi|Thriller  ...  English     USA   
25                               Drama|Romance  ...  English     USA   
2704           Action|Adventure|Fantasy|Sci-Fi  ...  English     USA   
2748                             Family|Sci-Fi  ...  English     USA   
16                     Action|Adventure|Sci-Fi  ...  English     USA   
482   Adventure|Animation|Drama|Family|Musical  ...  English     USA   
230            Action|Adventure|Fantasy|Sci-Fi  ...  English     USA   
64                 Action|Crime|Drama|Thriller  ...  English     USA   
419            Adventure|Drama|Sci-Fi|Thriller  ...  English     USA   

      content_rating       budget title_year  actor_2_facebook_likes  \
0              PG-13  237000000.0     2009.0                   936.0   
28             PG-13  150000000.0     2015.0                  2000.0   
25             PG-13  200000000.0     1997.0                 14000.0   
2704              PG   11000000.0     1977.0                  1000.0   
2748              PG   10500000.0     1982.0                   725.0   
16             PG-13  220000000.0     2012.0                 21000.0   
482                G   45000000.0     1994.0                   886.0   
230               PG  115000000.0     1999.0                 14000.0   
64             PG-13  185000000.0     2008.0                 13000.0   
419            PG-13   78000000.0     2012.0                 14000.0   

     imdb_score aspect_ratio  movie_facebook_likes   net_profit  
0           7.9         1.78                 33000  523505847.0  
28          7.0         2.00                150000  502177271.0  
25          7.7         2.35                 26000  458672302.0  
2704        8.7         2.35                 33000  449935665.0  
2748    

**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [109]:
# Write your code for creating the profit column here
# Write your code for unit conversion here
#movies.set_index('movie_title')
movies['net_profit'] = movies['gross'] - movies['budget']
movies.sort_values('net_profit', ascending = False)['movie_title'].head(10)

0                                          Avatar 
28                                 Jurassic World 
25                                        Titanic 
2704           Star Wars: Episode IV - A New Hope 
2748                   E.T. the Extra-Terrestrial 
16                                   The Avengers 
482                                 The Lion King 
230     Star Wars: Episode I - The Phantom Menace 
64                                The Dark Knight 
419                              The Hunger Games 
Name: movie_title, dtype: object

In [ ]:
# Write your code for sorting the dataframe here


In [ ]:
top10 = # Write your code to get the top 10 profiting movies here

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

**Subtask 3.3: Find IMDb Top 250**

Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000. 

Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [ ]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe 
# and name that dataframe as 'IMDb_Top_250'


#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_250*? </font>

**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [114]:
# Write your code for creating three new dataframes here
Meryl_Streep = movies[movies['actor_1_name'] == "Meryl Streep"] # Include all movies in which Meryl_Streep is the lead
#len(Meryl_Streep[['actor_1_name']])

11

In [127]:
Leo_Caprio = movies[(movies['actor_1_name'] == "Leonardo DiCaprio") ] # Include all movies in which Leo_Caprio is the lead
Leo_Caprio[['actor_1_name', 'actor_2_name', 'actor_3_name']]

actor_1_name         actor_2_name            actor_3_name
25    Leonardo DiCaprio         Kate Winslet           Gloria Stuart
49    Leonardo DiCaprio    Elizabeth Debicki            Steve Bisley
94    Leonardo DiCaprio            Tom Hardy    Joseph Gordon-Levitt
173   Leonardo DiCaprio            Tom Hardy              Lukas Haas
246   Leonardo DiCaprio           Adam Scott          Frances Conroy
283   Leonardo DiCaprio      Christoph Waltz            Ato Essandoh
293   Leonardo DiCaprio       Djimon Hounsou         Stephen Collins
294   Leonardo DiCaprio  Matthew McConaughey             Jon Favreau
312   Leonardo DiCaprio          Liam Neeson           Jim Broadbent
347   Leonardo DiCaprio           Matt Damon            Ray Winstone
431   Leonardo DiCaprio        Joseph Sikora          Nellie Sciutto
608   Leonardo DiCaprio       Simon McBurney          Michael Gaston
859   Leonardo DiCaprio            Tom Hanks         Jennifer Garner
934   Leonardo DiCaprio     Virginie Ledoyen  Peter Youngblood Hills
1046  Leonardo DiCaprio         Kate Winslet              Joe Komara
1336  Leonardo DiCaprio       Anne Parillaud         Judith Godrèche
1364  Leonardo DiCaprio          Naomi Watts           Kaitlyn Dever
1457  Leonardo DiCaprio           Pat Hingle         Roberts Blossom
1909  Leonardo DiCaprio       Robert De Niro            Meryl Streep
2484  Leonardo DiCaprio     Harold Perrineau           Brian Dennehy
3025  Leonardo DiCaprio    Elizabeth Debicki            Steve Bisley

In [128]:
Brad_Pitt = movies[movies['actor_1_name'] == "Brad Pitt"] # Include all movies in which Brad_Pitt is the lead
#Brad_Pitt[['actor_1_name', 'actor_2_name', 'actor_3_name']]

actor_1_name         actor_2_name         actor_3_name
97      Brad Pitt        Jason Flemyng         Julia Ormond
142     Brad Pitt        Orlando Bloom        Julian Glover
243     Brad Pitt        Julia Roberts           Mini Anden
244     Brad Pitt  Angelina Jolie Pitt      Stephanie March
367     Brad Pitt      Stephen Dillane  Catherine McCormack
383     Brad Pitt           Bernie Mac        Elliott Gould
448     Brad Pitt         Logan Lerman          Jim Parrack
579     Brad Pitt                 Mako          Victor Wong
646     Brad Pitt            Meat Loaf    Eugenie Bondurant
749     Brad Pitt     Adriano Giannini         Timothy West
887     Brad Pitt           Tom Cruise        Kirsten Dunst
1397    Brad Pitt         Tye Sheridan           Fiona Shaw
1608    Brad Pitt        Jeremy Renner          Sam Shepard
2030    Brad Pitt       Harriet Walter       Dermot Crowley
2146    Brad Pitt  Angelina Jolie Pitt       Melvil Poupaud
2428    Brad Pitt          Sam Shepard       Ben Mendelsohn
2602    Brad Pitt          Gary Oldman     Michael Rapaport

In [144]:
# Write your code for combining the three dataframes here
Combined = pd.concat([Meryl_Streep, Leo_Caprio, Brad_Pitt])
Combined

color          director_name  num_critic_for_reviews  \
392              Color           Nancy Meyers                   187.0   
1038             Color          Curtis Hanson                    42.0   
1132             Color            Nora Ephron                   252.0   
1322             Color          David Frankel                   208.0   
1390             Color         Robert Redford                   227.0   
1471             Color         Sydney Pollack                    66.0   
1514             Color          David Frankel                   234.0   
1563             Color          Carl Franklin                    64.0   
1784             Color         Stephen Daldry                   174.0   
2500             Color         Phyllida Lloyd                   331.0   
2793             Color          Robert Altman                   211.0   
25               Color          James Cameron                   315.0   
49               Color           Baz Luhrmann                   490.0   
94               Color      Christopher Nolan                   642.0   
173              Color  Alejandro G. Iñárritu                   556.0   
246    Black and White        Martin Scorsese                   267.0   
283              Color      Quentin Tarantino                   765.0   
293              Color           Edward Zwick                   166.0   
294              Color        Martin Scorsese                   606.0   
312              Color        Martin Scorsese                   233.0   
347              Color        Martin Scorsese                   352.0   
431              Color        Martin Scorsese                   490.0   
608              Color           Ridley Scott                   238.0   
859              Color       Steven Spielberg                   194.0   
934              Color            Danny Boyle                   118.0   
1046             Color             Sam Mendes                   323.0   
1336             Color        Randall Wallace                    83.0   
1364             Color         Clint Eastwood                   392.0   
1457   Black and White              Sam Raimi                    63.0   
1909             Color             Jerry Zaks                    45.0   
2484             Color           Baz Luhrmann                   106.0   
3025             Color           Baz Luhrmann                   490.0   
97               Color          David Fincher                   362.0   
142              Color      Wolfgang Petersen                   220.0   
243              Color      Steven Soderbergh                   198.0   
244              Color             Doug Liman                   233.0   
367              Color             Tony Scott                   142.0   
383              Color      Steven Soderbergh                   186.0   
448              Color             David Ayer                   406.0   
579              Color    Jean-Jacques Annaud                    76.0   
646              Color          David Fincher                   315.0   
749              Color        Patrick Gilmore                    98.0   
887              Color            Neil Jordan                   120.0   
1397             Color        Terrence Malick                   584.0   
1608             Color         Andrew Dominik                   273.0   
2030             Color  Alejandro G. Iñárritu                   285.0   
2146             Color    Angelina Jolie Pitt                   131.0   
2428             Color         Andrew Dominik                   414.0   
2602             Color             Tony Scott                   122.0   

      duration  director_facebook_likes  actor_3_facebook_likes  \
392      120.0                    278.0                   963.0   
1038     111.0                    161.0                   132.0   
1132     123.0                      0.0                   923.0   
1322     109.0                     64.0                   505.0   
1390      92.0                 

In [152]:
# Write your code for grouping the combined dataframe here
combined_actor = Combined.groupby('actor_1_name')
combined_actor['num_critic_for_reviews'].mean()

actor_1_name
Brad Pitt            245.000000
Leonardo DiCaprio    330.190476
Meryl Streep         181.454545
Name: num_critic_for_reviews, dtype: float64

In [ ]:
# Write the code for finding the mean of critic reviews and audience reviews here


#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>